src/Features/Core/Portable/FileHeaders/AbstractFileHeaderCodeFixProvider.cs

```diff
@@ -0,0 +1,243 @@
+
+            // Skip blank lines already at the beginning of the document, since we add our own
+            var leadingTrivia = root.GetLeadingTrivia();
+            var skipCount = 0;
+            for (var i = 0; i < leadingTrivia.Count; i++)
+            {
+                if (leadingTrivia[i].RawKind == SyntaxKinds.EndOfLineTrivia)
+                {
+                    skipCount = i + 1;
+                }
+                else if (leadingTrivia[i].RawKind != SyntaxKinds.WhitespaceTrivia)
+                {
+                    break;
+                }
+            }
+
+            newTrivia = newTrivia.AddRange(leadingTrivia.Skip(skipCount));
+
+            return root.WithLeadingTrivia(newTrivia);
+        }
+
+        private SyntaxTriviaList CreateNewHeader(string prefixWithLeadingSpaces, string expectedFileHeader, string newLineText)
+        {
+            var copyrightText = GetCopyrightText(prefixWithLeadingSpaces, expectedFileHeader, newLineText);
+            var newHeader = copyrightText;
+            return ParseLeadingTrivia(newHeader);
+        }
+
+        private static string GetCopyrightText(string prefixWithLeadingSpaces, string copyrightText, string newLineText)
+        {
+            copyrightText = copyrightText.Replace("\r\n", "\n");
+            var lines = copyrightText.Split('\n');
+            return string.Join(newLineText, lines.Select(line =>
+            {
+                if (string.IsNullOrEmpty(line))
+                {
+                    return prefixWithLeadingSpaces;
+                }
+                else
+                {
+                    return prefixWithLeadingSpaces + " " + line;
+                }
+            }));
+        }
+
+        private class MyCodeAction : CodeAction.DocumentChangeAction
+        {
+            public MyCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument)
+                : base(FeaturesResources.Add_file_banner, createChangedDocument, nameof(AbstractFileHeaderCodeFixProvider))
+            {
+            }
+        }
+
+        private class FixAll : DocumentBasedFixAllProvider
+        {
+            private readonly AbstractFileHeaderCodeFixProvider _codeFixProvider;
+
+            public FixAll(AbstractFileHeaderCodeFixProvider codeFixProvider)
+            {
+                _codeFixProvider = codeFixProvider;
+            }
+
+            protected override string CodeActionTitle => FeaturesResources.Add_file_banner;
+
+            protected override Task<SyntaxNode?> FixAllInDocumentAsync(FixAllContext fixAllContext, Document document, ImmutableArray<Diagnostic> diagnostics)
+            {
+                if (diagnostics.IsEmpty)
+                {
+                    return SpecializedTasks.Null<SyntaxNode>();
+                }
+
+                return _codeFixProvider.GetTransformedSyntaxRootAsync(document, fixAllContext.CancellationToken).AsNullable();
```

Comment:
ah... so this is why you want the SyntaxNode version from above :)

Well, now i reiterate that i would prefer `DocumentBasedFixAllProvider` (especially based on the name) to be *Document* based :D   i.e. it should be a `Document->Document` transform.  It would actually simplify this case as well.

Reply:
Fine with me.